{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Alphabet import generic_protein, generic_dna\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.SeqIO import to_dict, parse\n",
    "from os import listdir, mkdir\n",
    "from os.path import isdir\n",
    "\n",
    "def _sequence_back_translate(aligned_protein_record, unaligned_nucleotide_record, gap=None):\n",
    "    #TODO - Separate arguments for protein gap and nucleotide gap?\n",
    "    if hasattr(aligned_protein_record.seq.alphabet, \"gap_char\"):\n",
    "        if not gap:\n",
    "            gap = aligned_protein_record.seq.alphabet.gap_char\n",
    "        elif gap != aligned_protein_record.seq.alphabet.gap_char:\n",
    "            raise ValueError(\"Gap %r does not match %r from %r aligned protein alphabet\" \\\n",
    "                             % (gap, aligned_protein_record.seq.alphabet.gap_char,\n",
    "                                aligned_protein_record.id))\n",
    "    if not gap:\n",
    "        raise ValueError(\"Please supply the protein alignment gap character\")\n",
    "\n",
    "    alpha = unaligned_nucleotide_record.seq.alphabet\n",
    "    if hasattr(alpha, \"gap_char\"):\n",
    "        gap_codon = alpha.gap_char * 3\n",
    "        assert len(gap_codon) == 3\n",
    "    else:\n",
    "        from Bio.Alphabet import Gapped\n",
    "        alpha = Gapped(alpha, gap)\n",
    "        gap_codon = gap*3\n",
    "\n",
    "    seq = []\n",
    "    nuc = str(unaligned_nucleotide_record.seq)\n",
    "    for amino_acid in aligned_protein_record.seq:\n",
    "        if amino_acid == gap:\n",
    "            seq.append(gap_codon)\n",
    "        else:\n",
    "            seq.append(nuc[:3])\n",
    "            nuc = nuc[3:]\n",
    "    assert not nuc, \"Nucleotide sequence for %r longer than protein %s\" \\\n",
    "        % (unaligned_nucleotide_record.id, aligned_protein_record.id)\n",
    "\n",
    "    aligned_nuc = unaligned_nucleotide_record[:] #copy for most annotation\n",
    "    aligned_nuc.letter_annotation = {} #clear this\n",
    "    aligned_nuc.seq = Seq(\"\".join(seq), alpha) #replace this\n",
    "    if len(aligned_protein_record.seq) * 3 == len(aligned_nuc):\n",
    "        return aligned_nuc\n",
    "    else:\n",
    "        print('--->',len(aligned_protein_record.seq) * 3, len(aligned_nuc))\n",
    "        for i in range(len(aligned_protein_record.seq) * 3-len(aligned_nuc)):\n",
    "            aligned_nuc += '-'\n",
    "    return aligned_nuc\n",
    "\n",
    "def alignment_back_translate(protein_alignment, nucleotide_records, key_function=None, gap=None):\n",
    "    #TODO - Separate arguments for protein and nucleotide gap characters?\n",
    "    if key_function is None:\n",
    "        key_function = lambda x: x\n",
    "\n",
    "    if hasattr(protein_alignment._alphabet, \"gap_char\"):\n",
    "        if not gap:\n",
    "            gap = protein_alignment._alphabet.gap_char\n",
    "        elif gap != protein_alignment._alphabet.gap_char:\n",
    "            raise ValueError(\"Gap %r does not match %r from protein alignment alphabet\" \\\n",
    "                             % (gap, protein_alignment._alphabet.gap_char))\n",
    "\n",
    "    aligned = []\n",
    "    for protein in protein_alignment:\n",
    "        try:\n",
    "            nucleotide = nucleotide_records[key_function(protein.id)]\n",
    "        except KeyError:\n",
    "            print(str(protein.id))\n",
    "            raise ValueError(\"Could not find nucleotide sequence for protein %r\" % (protein.id))\n",
    "        temp = _sequence_back_translate(protein, nucleotide, gap)\n",
    "        aligned.append(temp)\n",
    "    A = set([])\n",
    "    for a in aligned:\n",
    "        A.add(len(a.seq))\n",
    "    print('Done')\n",
    "    print(A)\n",
    "    return MultipleSeqAlignment(aligned)\n",
    "\n",
    "def Parse_Nuc_Sequences(seq_path):\n",
    "    fasta_sequenes = parse(open(seq_path),'fasta')\n",
    "    sequences = []\n",
    "    try:\n",
    "        for f in fasta_sequenes:\n",
    "            temp = SeqRecord(Seq(str(f.seq)[:-2], generic_dna), id=str(f.name))\n",
    "            sequences.append(temp)\n",
    "    except UnicodeDecodeError:\n",
    "        pass\n",
    "    return to_dict(sequences)\n",
    "\n",
    "def Parse_MSA(seq_path):\n",
    "    fasta_sequenes = parse(open(seq_path),'fasta')\n",
    "    sequences = []\n",
    "    for f in fasta_sequenes:\n",
    "        temp = SeqRecord(Seq(str(f.seq), generic_protein), id=str(f.name))\n",
    "        sequences.append(temp)\n",
    "    return MultipleSeqAlignment(sequences)\n",
    "\n",
    "def Write_Nuc_MSA(nuc_alignments, op_filepath):\n",
    "    fbuf = open(op_filepath,'w')\n",
    "    for n in nuc_alignments:\n",
    "        fbuf.write('>'+str(n.id)+'\\n')\n",
    "        fbuf.write(str(n.seq)+'\\n')\n",
    "    fbuf.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_seq_path = '/Users/harihara/Mount/CMSC829A/Data/Accessory_Genes_Nucleotides/'\n",
    "dna_seq_files = listdir(dna_seq_path)\n",
    "prot_seq_path = '/Users/harihara/Mount/CMSC829A/Data/Accessory_Genes_Proteins_MSA/'\n",
    "out_path = '/Users/harihara/Mount/CMSC829A/Data/Accessory_Genes_Nucleotides_MSA/'\n",
    "if not isdir(out_path):\n",
    "    mkdir(out_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "Nucleotide sequence for 'GCF_000005845.2_ASM584v2.0' longer than protein GCF_000005845.2_ASM584v2.0",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-68-965ccc70ad54>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mdna_seqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mParse_Nuc_Sequences\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdna_seq_path\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mseq_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mprot_msa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mParse_MSA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprot_seq_path\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mseq_path\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\".fna\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\".faa\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0mnuc_msa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0malignment_back_translate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprot_msa\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdna_seqs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgap\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"-\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m     \u001b[0mWrite_Nuc_MSA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnuc_msa\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_path\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mseq_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mctr\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-66-e62a48a0af23>\u001b[0m in \u001b[0;36malignment_back_translate\u001b[0;34m(protein_alignment, nucleotide_records, key_function, gap)\u001b[0m\n\u001b[1;32m     69\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Could not find nucleotide sequence for protein %r\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 71\u001b[0;31m         \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_sequence_back_translate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprotein\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnucleotide\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgap\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     72\u001b[0m         \u001b[0maligned\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m     \u001b[0mA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-66-e62a48a0af23>\u001b[0m in \u001b[0;36m_sequence_back_translate\u001b[0;34m(aligned_protein_record, unaligned_nucleotide_record, gap)\u001b[0m\n\u001b[1;32m     37\u001b[0m             \u001b[0mnuc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnuc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0;32massert\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mnuc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Nucleotide sequence for %r longer than protein %s\"\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m         \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0munaligned_nucleotide_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maligned_protein_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0maligned_nuc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0munaligned_nucleotide_record\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m#copy for most annotation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAssertionError\u001b[0m: Nucleotide sequence for 'GCF_000005845.2_ASM584v2.0' longer than protein GCF_000005845.2_ASM584v2.0"
     ]
    }
   ],
   "source": [
    "ctr = 0\n",
    "dna_seq_files = ['copA.fna']\n",
    "for seq_path in dna_seq_files:\n",
    "    if(seq_path.startswith(\".\")):\n",
    "        continue\n",
    "    \n",
    "    dna_seqs = Parse_Nuc_Sequences(dna_seq_path+seq_path)\n",
    "    prot_msa = Parse_MSA(prot_seq_path+seq_path.replace(\".fna\",\".faa\"))\n",
    "    nuc_msa = alignment_back_translate(prot_msa, dna_seqs,gap = \"-\")\n",
    "    Write_Nuc_MSA(nuc_msa, out_path+seq_path)\n",
    "    print(ctr,seq_path)\n",
    "    #except AssertionError:\n",
    "    #    print('--->',ctr,seq_path)\n",
    "    #except ValueError:\n",
    "    #    print('+++>',ctr,seq_path)\n",
    "    ctr += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
