{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sys import exit\n",
    "from os.path import isdir, isfile\n",
    "from os import listdir, mkdir\n",
    "import Bio.SeqIO as IO\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Parse_HMM_Alignments(align):\n",
    "    OP = []\n",
    "    counts = 0\n",
    "    for a in align.split(\"//\\n\"):\n",
    "        o = a.split(\"\\n\")\n",
    "        align_flag = False\n",
    "        ctr = 0\n",
    "        while ctr < len(o):\n",
    "            if o[ctr].lstrip().startswith(\"Query:\"):\n",
    "                query = o[ctr].lstrip().replace(\"Query:\",\"\").lstrip().split(' ')[0]\n",
    "                ctr += 1\n",
    "            elif o[ctr].lstrip().startswith('>>'):\n",
    "                subject = o[ctr].lstrip().replace(\">>\",\"\").rstrip()\n",
    "                aligns = []\n",
    "                ctr += 2\n",
    "                while (\"Alignments for each domain\" not in o[ctr]) and (\"Internal pipeline statistics\" not in o[ctr]):\n",
    "                    if o[ctr].lstrip().startswith(\"--\"):\n",
    "                        ctr += 1\n",
    "                        continue\n",
    "                    values = o[ctr].lstrip().replace(\"[.\",\"\").replace(\".]\",\"\").replace(\"..\",\"\").replace(\"[]\",\"\").split(' ')\n",
    "                    values = [t for t in values if t != '']\n",
    "                    ctr += 1\n",
    "                    header = ['No', 'Score','bias','c-Evalue', 'i-Evalue','hmmfrom','hmmto',\n",
    "                              'alifrom','alito','envfrom', 'envto', 'acc']\n",
    "                    values = [float(v) for v in values if v != '!' and v != '?']\n",
    "                    d = dict(zip(header, values))\n",
    "                    d['Query'] = query\n",
    "                    d['Subject'] = subject\n",
    "                    d['Subject_Alignment'] = ''\n",
    "                    d['Alignment_Description'] = ''\n",
    "                    d['Query_Alignment'] = ''\n",
    "                    aligns.append(d)\n",
    "                \n",
    "                if o[ctr].lstrip().startswith('Alignments for each domain') :\n",
    "                    cur = 0\n",
    "                    align_flag = True\n",
    "                    ctr += 2\n",
    "            \n",
    "            if(align_flag):\n",
    "                if (o[ctr].lstrip().startswith(\"==\")):\n",
    "                    cur += 1\n",
    "                    ctr += 1\n",
    "                tgt = o[ctr].lstrip().split(' ')\n",
    "                tgt = [t for t in tgt if t != '']\n",
    "                mat = o[ctr+1].lstrip().replace(\" \",\"S\")\n",
    "                qry = o[ctr+2].lstrip().split(' ')\n",
    "                qry = [t for t in qry if t != '']\n",
    "                aligns[cur]['Subject_Alignment'] += tgt[2]\n",
    "                aligns[cur]['Query_Alignment']  += qry[2]\n",
    "                aligns[cur]['Alignment_Description']  += 'S'*(len(mat)-len(tgt[2]))+mat\n",
    "                ctr += 4\n",
    "\n",
    "                if o[ctr].startswith('Internal') or o[ctr].lstrip().startswith('>>') :\n",
    "                    counts += 1\n",
    "                    OP += aligns\n",
    "                    align_flag = False\n",
    "                if o[ctr].startswith('Internal'):\n",
    "                    break\n",
    "            else:\n",
    "                ctr += 1\n",
    "            \n",
    "            \n",
    "    df_OP = pd.DataFrame(OP)\n",
    "    return df_OP\n",
    "\n",
    "def Calculate_Untagged_Gene_Lengths(filepath):\n",
    "    d = {}\n",
    "    record_dict = IO.to_dict(IO.parse(filepath, \"fasta\"))\n",
    "    for key in record_dict.items():\n",
    "        d[key[0]] = len(key[1].seq)\n",
    "    df_lengths = pd.DataFrame(data = {'Query':list(d.keys()), 'Length':list(d.values())})\n",
    "    return df_lengths\n",
    "\n",
    "def Load_HMMER_Outputs(filepath, df_lengths):\n",
    "    df = pd.read_csv(filepath)\n",
    "    del df['Unnamed: 0']\n",
    "   \n",
    "    return df\n",
    "\n",
    "def Return_Breadth_of_Coverage(grp):\n",
    "    grp = grp.sort_values(by = 'alifrom')\n",
    "    starts = grp['alifrom'].tolist()\n",
    "    ends = grp['alito'].tolist()\n",
    "    breadth = 0\n",
    "    for i in range(1, len(starts)):\n",
    "        prev_end = ends[i-1]\n",
    "        curr_starts = starts[i]\n",
    "        dist = curr_starts - prev_end\n",
    "        if dist > 0:\n",
    "            breadth += dist\n",
    "    breadth = (starts[0] - 0) + breadth + (grp.iloc[0]['Length'] - ends[-1])\n",
    "    breadth = breadth/grp.iloc[0]['Length'] * 100.0\n",
    "    norm = np.sum(grp['alito'] - grp['alifrom'])\n",
    "    score = np.sum((grp['alito'] - grp['alifrom'])*grp['Score']/(norm))\n",
    "    cevlaue = np.sum((grp['alito'] - grp['alifrom'])*grp['c-Evalue']/norm)\n",
    "    ievlaue = np.sum((grp['alito'] - grp['alifrom'])*grp['i-Evalue']/norm)\n",
    "    \n",
    "    return pd.Series({'breadth':breadth, 'Counts':len(grp), 'Avg. Score':score, \n",
    "                      'Avg. c-Evalue':cevlaue, 'Avg. i-Evalue':ievlaue})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_019048845.1_ASM1904884v1_genomic.out\n"
     ]
    }
   ],
   "source": [
    "filepath = '/Users/harihara/UMD_Coursework/Fall_2021/CMSC829A/Project/Data/Untagged_Accessory_Genes_Proteins_HMM_Alignment/'\n",
    "outdir = '/Users/harihara/Mount/CMSC829A/Data/Untagged_Accessory_Genes_Proteins_HMM_Alignment_Processed/'\n",
    "tophit_dir = '/Users/harihara/Mount/CMSC829A/Data/Untagged_Accessory_Genes_Proteins_HMM_Alignments_Tophits/'\n",
    "untagged_gene_seq = '/Users/harihara/Mount/CMSC829A/Data/Untagged_Accessory_Genes_Proteins/'\n",
    "\n",
    "if not isdir(tophit_dir):\n",
    "    mkdir(tophit_dir)\n",
    "files = listdir(filepath)\n",
    "\n",
    "for f in files:\n",
    "    #print(f)\n",
    "    #if isfile(outdir+f):\n",
    "    #    continue\n",
    "    try:\n",
    "        if f.startswith(\"GCF\") and f.endswith(\".out\"):\n",
    "            lines = open(filepath+f).readlines()\n",
    "            out = ''\n",
    "            for line in lines:\n",
    "                if line.startswith('#') or line.startswith('--') or line == \"\\n\":\n",
    "                    continue\n",
    "                else:\n",
    "                    out += line\n",
    "            df_HMMER = Parse_HMM_Alignments(out)\n",
    "            df_lengths = Calculate_Untagged_Gene_Lengths(untagged_gene_seq+f.replace(\".out\",\".faa\"))\n",
    "            df_HMMER = pd.merge(df_HMMER, df_lengths, on = ['Query'], how = 'left')\n",
    "            df_HMMER = df_HMMER[df_HMMER['Score'] >= 0]\n",
    "            df_HMMER_grp = df_HMMER.groupby(['Query','Subject']).apply(Return_Breadth_of_Coverage)\n",
    "            df_seqs = df_HMMER[['Query','Subject','Length','Subject_Alignment','Alignment_Description','Query_Alignment']]\n",
    "            df_seqs = df_seqs.set_index(['Query', 'Subject'])\n",
    "            df_HMMER_grp = df_HMMER_grp.join(df_seqs, how = 'left')\n",
    "            df_HMMER_grp = df_HMMER_grp.reset_index()\n",
    "            min_breadth = 60\n",
    "            df_HMMER_filter = df_HMMER_grp[df_HMMER_grp['breadth'] >= min_breadth]\n",
    "            df_HMMER_filter = df_HMMER_filter.loc[df_HMMER_filter.groupby('Query')['Avg. Score'].idxmax()]\n",
    "            df_HMMER_filter.to_csv(outdir+f.replace(\".faa\",\".txt\"), sep = '\\t')\n",
    "            df_HMMER_filter.to_csv(tophit_dir+f)\n",
    "            print(f)\n",
    "            break\n",
    "    except IndexError:\n",
    "        print('---->', f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
